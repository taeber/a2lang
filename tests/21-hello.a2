asm {*
	ORG $800
DOS	EQU $3D0
	JSR main
	JMP DOS
}

use [
    INIT: sub @ $FB2F,  ; AKA TEXT
    HOME: sub @ $FC58,

    COUT : sub <- [ch: char @ A] @ $FDED,
    CROUT: sub @ $FD8E,
]

let Println = sub <- [txt: text @ $06] {
    var [
        i  :int @ Y,
        ch :char,
    ]

    i := 0
    loop if txt_i <> 0 {
        COUT(txt_i)
        i  += 1
    }

    CROUT()
}

use PRTLN: sub <- [txt: text @ $06]

let main = sub {
    INIT()
    HOME()
    Println("Hello, 1979!")
    PRTLN("Hello, 1979!")
}


asm {
PRTLN	LDY #0
.loop	LDA (.txt),Y
	BEQ .done
	JSR COUT
	INY
	JMP .loop
.done	JSR CROUT
	RTS
}
