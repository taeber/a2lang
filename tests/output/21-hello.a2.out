*
	ORG $800
DOS	EQU $3D0
	JSR main
	JMP DOS
INIT	EQU $FB2F
HOME	EQU $FC58
COUT	EQU $FDED
CROUT	EQU $FD8E
Println.txt	EQU $06
* COPYBB @Y #$00
Println	LDY #$00
* IFNE (Println.txt),@Y #$00
*   Println._1 Println._2
Println._0	LDA (Println.txt),Y
	BNE Println._1
	JMP Println._2
* COPYBB @A (Println.txt),@Y
Println._1	LDA (Println.txt),Y
	JSR COUT
* ADDBB @Y #$01
	INY
	JMP Println._0
Println._2	JSR CROUT
	RTS
PRTLN.txt	EQU $06
main	JSR INIT
	JSR HOME
	LDA #<A2_3
	LDX #>A2_3
	STX Println.txt+1
	STA Println.txt
	JSR Println
	LDA #<A2_4
	LDX #>A2_4
	STX PRTLN.txt+1
	STA PRTLN.txt
	JMP PRTLN
PRTLN	LDY #0
.loop	LDA (.txt),Y
	BEQ .done
	JSR COUT
	INY
	JMP .loop
.done	JSR CROUT
	RTS
Println.ch	HEX 00
A2_3	ASC "Hello, 1979!"
	HEX 00
A2_4	ASC "Hello, 1979!"
	HEX 00
